Unit Tests for Enhanced Program Management Tool Doctype Module

SymonMuchemi opened the pull request:
This pull request introduces comprehensive test coverage for the `Enhanced Program Enrollment Tool` in `nl_school/junior_school_customization`. It includes unit tests for various scenarios, ensuring robust validation, correct functionality, and proper handling of edge cases.

### New Test Cases for `Enhanced Program Enrollment Tool`:

#### Core Functionality Testing:
* Added `test_get_students_from_applicant` to validate that students are correctly retrieved from approved `Student Applicants` based on the specified program, academic year, and term.
* Added `test_get_students_from_program_enrollment` to ensure students are retrieved accurately from `Program Enrollment` records for a given academic year and term.

#### Validation and Error Handling:
* Added `test_get_students_returns_error_when_no_students_found` to confirm an error is raised when no students match the specified criteria.
* Added `test_get_students_throws_if_academic_year_missing` and `test_get_students_throws_if_program_missing_for_applicant` to validate that mandatory fields (`Academic Year` and `Program`) are enforced.
* Added `test_get_students_with_invalid_source_throws_no_students` to ensure an error is raised when an invalid source is provided.

#### Edge Case Handling
This pull request enhances the test coverage for the `Enhanced Program Enrollment Tool` by adding multiple test cases to validate its functionality. The changes include setting up and tearing down test data, testing various scenarios for retrieving students, and verifying error handling and workflows.

### Test Coverage Enhancements:

* **Setup and Teardown for Test Data**:
  - Added `setUp` and `tearDown` methods to create and clean up test data for `Student Applicant`, `Student`, and `Program Enrollment` records. This ensures consistent test environments.

* **New Test Cases for `get_students` Method**:
  - Added tests to validate that students are correctly retrieved from `Student Applicant` and `Program Enrollment` based on specified criteria.
  - Implemented a test to ensure disabled students are filtered out when retrieving students from `Program Enrollment`.
  - Added tests to verify that `get_students` raises appropriate errors when no students are found, or when mandatory fields like `academic_year` or `program` are missing.
  - Included a test to handle invalid sources for the `get_students_from` field, ensuring proper error messages are raised.

* **Workflow Validation for `enroll_students` Method
This pull request adds extensive test coverage for the `Enhanced Program Enrollment Tool` in the `nl_school` module. The changes include new test cases to validate the tool's functionality, error handling, and workflows. Additionally, the `setUp` and `tearDown` methods are implemented to manage test data setup and cleanup.

### New Test Cases for `Enhanced Program Enrollment Tool`:

* **Basic Functionality:**
  - Added `test_get_students_from_applicant` to verify that students are correctly retrieved from approved `Student Applicant` records.
  - Added `test_get_students_from_program_enrollment` to validate that students are retrieved from `Program Enrollment` records for a given academic year and term.

* **Error Handling:**
  - Added `test_get_students_returns_error_when_no_students_found` to confirm that an error is raised when no students match the specified criteria.
  - Added `test_get_students_throws_if_academic_year_missing` and `test_get_students_throws_if_program_missing_for_applicant` to ensure that mandatory fields (`Academic Year` and `Program`) are validated.
  - Added `test_get_students_with_invalid_source_throws_no_students` to validate error handling for invalid data sources.

SymonMuchemi committed:
feat: Implement setup and teardown methods for Enhanced Program Enrollment Tool tests

Added setup and teardown methods in the TestEnhancedProgramEnrollmentTool class to create and clean up test data for Student Applicant, Student, and Program Enrollment. This includes inserting necessary documents into the database for testing purposes and ensuring proper cleanup after tests are run.

SymonMuchemi committed:
feat: Enhance test setup for Enhanced Program Enrollment Tool

- Refactored the setup method to improve readability by formatting the insertion of test data for Student Applicant, Student, and Program Enrollment.
- Added a new test method to verify the retrieval of students from approved Student Applicants based on the specified program, academic year, and term.
- Ensured that the test checks if the created applicant is included in the results from the get_students method.

SymonMuchemi committed:
feat: Add test for retrieving students from Program Enrollment

This commit introduces a new test case to the Enhanced Program Enrollment Tool that verifies the retrieval of students enrolled via Program Enrollment for a specified academic year and term. The test ensures that the tool correctly identifies and returns the expected student based on the provided enrollment details.

SymonMuchemi committed:
test: Add test for filtering out disabled students in Program Enrollment

This commit introduces a new test case to ensure that disabled students are not included in the results when retrieving students from Program Enrollment. The test sets up a disabled student and verifies that they are correctly filtered out by the Enhanced Program Enrollment Tool when fetching students for a specified academic year and term.

SymonMuchemi committed:
test: Add error handling test for no students found scenario

Added a test to verify that the get_students method raises a ValidationError when no students are found for the specified criteria. This includes creating a tool instance with filters that match no records and confirming that the appropriate exception is raised with the expected message.

SymonMuchemi committed:
feat: Add workflow trigger for enrolling students in Enhanced Program Enrollment Tool

This commit introduces a new test case that verifies the expected workflow when enrolling students through the Enhanced Program Enrollment Tool. The test mocks dependencies and checks that the enrollment process correctly triggers the associated functions and displays the appropriate message upon successful enrollment of students.

SymonMuchemi committed:
test: Add validation tests for student retrieval methods in Enhanced Program Enrollment Tool

- Implemented tests to ensure that the `get_students` method raises appropriate validation errors when required fields are missing.
- Added a test for missing academic year, which checks for a `ValidationError` with the message "Mandatory field - Academic Year".
- Added a test for missing program, which checks for a `ValidationError` with the message "Mandatory field - Program".
- Included a test for invalid source input, ensuring that it raises a `ValidationError` with the message "No students Found".
- Enhanced the existing test suite to improve coverage and reliability of the Enhanced Program Enrollment Tool functionality.

maniamartial commented on nl_school/junior_school_customization/doctype/enhanced_program_enrollment_tool/test_enhanced_program_enrollment_tool.py:
```diff
@@ -1,9 +1,271 @@
+            }
+        ).insert(ignore_permissions=True)
+
+    def tearDown(self):
```
After teardown(), which apparently will delete the defined doctype, commit because it's the last function being called.
frappe.db.commit()

SymonMuchemi replied:
Suggested change in this commit [00f29e7](https://github.com/navariltd/Junior-School/pull/73/commits/00f29e783cddf6772f8d6f21bc91823a29068d1d)

maniamartial commented on nl_school/junior_school_customization/doctype/enhanced_program_enrollment_tool/test_enhanced_program_enrollment_tool.py:
```diff
@@ -1,9 +1,271 @@
+        self.academic_term = "Term 1"
+
+        # Insert a Student Applicant
+        self.student_applicant = frappe.get_doc(
```
Before inserting/creating the test doc, please make sure you validate that they don't exist.
`if frappe.db.exists("Student Applicant", {values}, cache=False):
            return.`

SymonMuchemi replied:
Suggested changes in commits [091946c](https://github.com/navariltd/Junior-School/pull/73/commits/091946c5071762e7b6496e208b69f9a81dabfddc) and [1d49981](https://github.com/navariltd/Junior-School/pull/73/commits/1d49981ad07a66be3aeebb7ad9bf413fde68fe15)

SymonMuchemi committed:
fix: Ensure database changes are committed after teardown in tests

Added a commit statement in the tearDown method of the TestEnhancedProgramEnrollmentTool class to ensure that all changes made during the tests are properly committed to the database.

SymonMuchemi committed:
fix: Ensure Student Applicant is only created if it does not exist

Modified the setup method to check for the existence of a Student Applicant before creating a new one. If an applicant already exists, it retrieves the existing document instead of creating a duplicate.

SymonMuchemi committed:
update: disable caching for test student applicant query

SymonMuchemi committed:
update: Refactor student applicant and program enrollment creation logic

- Simplified the logic for inserting a Student Applicant by using the new get_or_create_doc helper method, ensuring that a Student Applicant is only created if it does not already exist.
- Updated the creation of approved Student Applicants in the test cases to utilize the same helper method for consistency and clarity.
- Refactored the Program Enrollment creation to also use the get_or_create_doc method, streamlining the process and reducing code duplication.
